sdk/metric/periodic_reader.go

```diff
@@ -0,0 +1,226 @@
+// that exceed 30 seconds. The export time is not counted towards the interval
+// between attempts.
+//
+// The Collect method of the returned Reader continues to gather and return
+// metric data to the user. It will not automatically send that data to the
+// exporter. That is left to the user to accomplish.
+func NewPeriodicReader(exporter Exporter, options ...PeriodicReaderOption) Reader {
+	conf := newPeriodicReaderConfig(options)
+	ctx, cancel := context.WithCancel(context.Background())
+	r := &periodicReader{
+		timeout:  conf.timeout,
+		exporter: exporter,
+		cancel:   cancel,
+	}
+
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		r.run(ctx, conf.interval)
+	}()
+
+	return r
+}
+
+// periodicReader is a Reader that continuously collects and exports metric
+// data at a set interval.
+type periodicReader struct {
+	producer atomic.Value
+
+	timeout  time.Duration
+	exporter Exporter
+
+	wg           sync.WaitGroup
+	cancel       context.CancelFunc
+	shutdownOnce sync.Once
+}
+
+// newTicker allows testing override.
+var newTicker = time.NewTicker
+
+// run continuously collects and exports metric data at the specified
+// interval. This will run until ctx is canceled or times out.
+func (r *periodicReader) run(ctx context.Context, interval time.Duration) {
+	ticker := newTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			m, err := r.Collect(ctx)
+			if err == nil {
+				c, cancel := context.WithTimeout(ctx, r.timeout)
+				err = r.exporter.Export(c, m)
+				cancel()
+			}
+			if err != nil {
+				otel.Handle(err)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// register registers p as the producer of this reader.
+func (r *periodicReader) register(p producer) {
+	r.producer.Store(produceHolder{produce: p.produce})
+}
+
+// Collect gathers and returns all metric data related to the Reader from
+// the SDK. The returned metric data is not exported to the configured
+// exporter, it is left to the caller to handle that if desired.
+//
+// An error is returned if this is called after Shutdown.
+func (r *periodicReader) Collect(ctx context.Context) (export.Metrics, error) {
+	p := r.producer.Load()
+	if p == nil {
+		return export.Metrics{}, ErrReaderNotRegistered
+	}
+
+	ph, ok := p.(produceHolder)
+	if !ok {
+		// The atomic.Value is entirely in the periodicReader's control so
+		// this should never happen. In the unforeseen case that this does
+		// happen, return an error instead of panicking so a users code does
+		// not halt in the processes.
+		err := fmt.Errorf("periodic reader: invalid producer: %T", p)
+		return export.Metrics{}, err
+	}
+	return ph.produce(ctx)
+}
+
+// ForceFlush flushes the Exporter.
+func (r *periodicReader) ForceFlush(ctx context.Context) error {
+	return r.exporter.ForceFlush(ctx)
+}
+
+// Shutdown stops the export pipeline.
+func (r *periodicReader) Shutdown(ctx context.Context) error {
```

Comment:
If this is called before the reader is registered, a misuse for sure, the ticker will be stopped, but the Collect method will still work.

I used this test to verify this:
```go
    t.Run("Shutdown Then register", func(t *testing.T) {
		r := f()

		err := r.Shutdown(context.Background())
		require.NoError(t, err)

		r.register(testProducer{})

		m, err := r.Collect(context.Background())
		assert.ErrorIs(t, err, ErrReaderShutdown)
		assert.Equal(t, export.Metrics{}, m)
	})
```

Reply:
Ah, good point. I'll add this test and include a fix. :+1: 

This also reminds me about [this part of the specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#metricreader):

> The SDK MUST NOT allow a MetricReader instance to be registered on more than one MeterProvider instance.

I'll be sure to add a test to validate that as well.

Reply:
Duplicate registration check: https://github.com/open-telemetry/opentelemetry-go/pull/2914